File: java/test/src/main/java/test/Ice/background/Acceptor.java
// Copyright (c) ZeroC, Inc.

package test.Ice.background;

import com.zeroc.Ice.ReadyCallback;

import java.nio.channels.ServerSocketChannel;

class Acceptor implements com.zeroc.Ice.Acceptor {
    @Override
    public ServerSocketChannel fd() {
        return _acceptor.fd();
    }

    @Override
    public void setReadyCallback(ReadyCallback callback) {
        // No need to for the ready callback.
    }

    @Override
    public void close() {
        _acceptor.close();
    }

    @Override
    public com.zeroc.Ice.EndpointI listen() {
        _endpoint = _endpoint.endpoint(_acceptor.listen());
        return _endpoint;
    }

    @Override
    public com.zeroc.Ice.Transceiver accept() {
        return new Transceiver(_configuration, _acceptor.accept());
    }

    @Override
    public String protocol() {
        return _acceptor.protocol();
    }

    @Override
    public String toString() {
        return _acceptor.toString();
    }

    @Override
    public String toDetailedString() {
        return _acceptor.toDetailedString();
    }

    public com.zeroc.Ice.Acceptor delegate() {
        return _acceptor;
    }

    Acceptor(EndpointI endpoint, Configuration configuration, com.zeroc.Ice.Acceptor acceptor) {
        _endpoint = endpoint;
        _configuration = configuration;
        _acceptor = acceptor;
    }

    private EndpointI _endpoint;
    private final com.zeroc.Ice.Acceptor _acceptor;
    private final Configuration _configuration;
}

File: csharp/src/IceSSL/Acceptor.cs
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

namespace IceSSL
{
    internal class Acceptor : IceInternal.IAcceptor
    {
        public void Close() => _delegate.Close();

        public Ice.Endpoint Listen()
        {
            _endpoint = _endpoint.GetEndpoint(_delegate.Listen());
            return _endpoint;
        }

        public bool StartAccept(IceInternal.AsyncCallback callback, object state)
        {
            //
            // The plug-in may not be fully initialized.
            //
            if (!_instance.Initialized())
            {
                throw new System.InvalidOperationException("IceSSL: plug-in is not initialized");
            }
            return _delegate.StartAccept(callback, state);
        }

        public void FinishAccept() => _delegate.FinishAccept();

        public IceInternal.ITransceiver Accept() => new Transceiver(_instance, _delegate.Accept(), _adapterName, true);

        public string Transport() => _delegate.Transport();

        public override string ToString() => _delegate.ToString();

        public string ToDetailedString() => _delegate.ToDetailedString();

        internal Acceptor(Endpoint endpoint, Instance instance, IceInternal.IAcceptor del, string adapterName)
        {
            _endpoint = endpoint;
            _delegate = del;
            _instance = instance;
            _adapterName = adapterName;

            //
            // .NET requires that a certificate be supplied.
            //
            System.Security.Cryptography.X509Certificates.X509Certificate2Collection? certs = instance.Certs();
            if (certs == null || certs.Count == 0)
            {
                throw new Ice.SecurityException("IceSSL: certificate required for server endpoint");
            }
        }

        private Endpoint _endpoint;
        private readonly IceInternal.IAcceptor _delegate;
        private readonly Instance _instance;
        private readonly string _adapterName;
    }
}
